### ISSUE TITLE
MPS backend produces bad training results in comparison to other backends

### ISSUE THREAD

**Issue by paulkre** [pytorch/pytorch#92615]
### 🐛 Describe the bug

Using the [MPS backend](https://pytorch.org/docs/stable/notes/mps.html) to train a model produces much worse results than using other backends (e.g. CPU or CUDA). To be clear, I am not talking about the speed of the training, but rather about the metrics for the quality (loss, perplexity) of the model after it has been trained. For example, if you run the training of the [word_language_model](https://github.com/pytorch/examples/tree/40289773aa4916fad0d50967917b3ae8aa534fd6/word_language_model) in the [pytorch/examples](https://github.com/pytorch/examples/tree/40289773aa4916fad0d50967917b3ae8aa534fd6) repository with either the CPU or the CUDA backend, your values for `loss` and `ppl` will be similar to this:

```
$ python main.py --cuda --epochs 1

| epoch   1 |   200/ 2983 batches | lr 20.00 | ms/batch 18.51 | loss  7.63 | ppl  2063.29
| epoch   1 |   400/ 2983 batches | lr 20.00 | ms/batch 17.47 | loss  6.86 | ppl   950.96
| epoch   1 |   600/ 2983 batches | lr 20.00 | ms/batch 17.47 | loss  6.48 | ppl   653.41
| epoch   1 |   800/ 2983 batches | lr 20.00 | ms/batch 17.46 | loss  6.29 | ppl   539.53
| epoch   1 |  1000/ 2983 batches | lr 20.00 | ms/batch 17.50 | loss  6.14 | ppl   465.56
| epoch   1 |  1200/ 2983 batches | lr 20.00 | ms/batch 17.54 | loss  6.07 | ppl   430.74
| epoch   1 |  1400/ 2983 batches | lr 20.00 | ms/batch 17.52 | loss  5.95 | ppl   384.86
| epoch   1 |  1600/ 2983 batches | lr 20.00 | ms/batch 17.55 | loss  5.96 | ppl   387.05
| epoch   1 |  1800/ 2983 batches | lr 20.00 | ms/batch 17.54 | loss  5.82 | ppl   337.38
| epoch   1 |  2000/ 2983 batches | lr 20.00 | ms/batch 17.52 | loss  5.80 | ppl   329.33
| epoch   1 |  2200/ 2983 batches | lr 20.00 | ms/batch 17.54 | loss  5.67 | ppl   289.19
| epoch   1 |  2400/ 2983 batches | lr 20.00 | ms/batch 17.53 | loss  5.67 | ppl   290.80
| epoch   1 |  2600/ 2983 batches | lr 20.00 | ms/batch 17.53 | loss  5.66 | ppl   285.86
| epoch   1 |  2800/ 2983 batches | lr 20.00 | ms/batch 17.54 | loss  5.55 | ppl   256.72
-----------------------------------------------------------------------------------------
| end of epoch   1 | time: 54.77s | valid loss  5.53 | valid ppl   252.11
-----------------------------------------------------------------------------------------
=========================================================================================
| End of training | test loss  5.44 | test ppl   230.14
=========================================================================================
```

Running the same training with MPS enabled consistently results in significantly worse values for `loss` and `ppl` (tested with pytorch `v1.13.1` and `v2.0.0.dev20230119`):

```
$ python main.py --mps --epochs 1

| epoch   1 |   200/ 2983 batches | lr 20.00 | ms/batch 115.26 | loss  7.99 | ppl  2959.03
| epoch   1 |   400/ 2983 batches | lr 20.00 | ms/batch 114.31 | loss  7.52 | ppl  1849.67
| epoch   1 |   600/ 2983 batches | lr 20.00 | ms/batch 114.39 | loss  7.38 | ppl  1603.87
| epoch   1 |   800/ 2983 batches | lr 20.00 | ms/batch 113.73 | loss  7.30 | ppl  1475.25
| epoch   1 |  1000/ 2983 batches | lr 20.00 | ms/batch 113.39 | loss  7.26 | ppl  1421.42
| epoch   1 |  1200/ 2983 batches | lr 20.00 | ms/batch 113.48 | loss  7.25 | ppl  1406.03
| epoch   1 |  1400/ 2983 batches | lr 20.00 | ms/batch 113.49 | loss  7.18 | ppl  1317.59
| epoch   1 |  1600/ 2983 batches | lr 20.00 | ms/batch 113.44 | loss  7.19 | ppl  1330.15
| epoch   1 |  1800/ 2983 batches | lr 20.00 | ms/batch 114.70 | loss  7.16 | ppl  1280.67
| epoch   1 |  2000/ 2983 batches | lr 20.00 | ms/batch 113.67 | loss  7.16 | ppl  1285.51
| epoch   1 |  2200/ 2983 batches | lr 20.00 | ms/batch 113.82 | loss  7.16 | ppl  1288.31
| epoch   1 |  2400/ 2983 batches | lr 20.00 | ms/batch 113.25 | loss  7.13 | ppl  1247.44
| epoch   1 |  2600/ 2983 batches | lr 20.00 | ms/batch 113.25 | loss  7.18 | ppl  1313.82
| epoch   1 |  2800/ 2983 batches | lr 20.00 | ms/batch 113.55 | loss  7.17 | ppl  1301.76
-----------------------------------------------------------------------------------------
| end of epoch   1 | time: 362.77s | valid loss  7.01 | valid ppl  1112.40
-----------------------------------------------------------------------------------------
=========================================================================================
| End of training | test loss  6.93 | test ppl  1019.79
=========================================================================================
```

Increasing the number of epochs will only increase the difference between the results of the two training configurations.

### Versions

**Latest:**

```
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.10.9 (main, Jan 11 2023, 09:18:18) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[conda] numpy                     1.23.5          py310hb93e574_0
[conda] numpy-base                1.23.5          py310haf87e8b_0
[conda] pytorch                   1.13.1                 py3.10_0    pytorch
```

**Nightly:**

```
PyTorch version: 2.0.0.dev20230119
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.10.9 (main, Jan 11 2023, 09:18:18) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0.dev20230119
[conda] numpy                     1.23.5          py310hb93e574_0
[conda] numpy-base                1.23.5          py310haf87e8b_0
[conda] pytorch                   2.0.0.dev20230119        py3.10_0    pytorch-nightly
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by qqaatw** [comment]
Hello, the bad results come from a numerical correctness issue of LSTM on MPS backend. We're working on fixing this.

**Comment by paulkre** [comment]
Is there an ETA for the next release?

**Comment by kulinseth** [comment]
> Is there an ETA for the next release?

It should be part of the 2.0 release.

**Comment by rgb1380** [comment]
I am still seeing a very similar issue with 2.0 when LTSM is used. Can someone confirm that the issue described here has been definitely fixed?

**Comment by qqaatw** [comment]
@rgb1380 Some details: https://github.com/pytorch/pytorch/issues/90421

TL;DR, please try PyTorch nightly if there are issues with 2.0.

**Comment by paulkre** [comment]
For me, this specific issue is fixed on 2.0:

```
$ python main.py --mps --epochs 1

| epoch   1 |   200/ 2983 batches | lr 20.00 | ms/batch 100.50 | loss  7.68 | ppl  2159.21
| epoch   1 |   400/ 2983 batches | lr 20.00 | ms/batch 99.01 | loss  6.89 | ppl   981.78
| epoch   1 |   600/ 2983 batches | lr 20.00 | ms/batch 99.95 | loss  6.56 | ppl   707.21
| epoch   1 |   800/ 2983 batches | lr 20.00 | ms/batch 98.91 | loss  6.39 | ppl   598.05
| epoch   1 |  1000/ 2983 batches | lr 20.00 | ms/batch 98.94 | loss  6.24 | ppl   513.87
| epoch   1 |  1200/ 2983 batches | lr 20.00 | ms/batch 98.71 | loss  6.15 | ppl   466.75
| epoch   1 |  1400/ 2983 batches | lr 20.00 | ms/batch 98.92 | loss  6.03 | ppl   414.00
| epoch   1 |  1600/ 2983 batches | lr 20.00 | ms/batch 98.98 | loss  6.02 | ppl   411.47
| epoch   1 |  1800/ 2983 batches | lr 20.00 | ms/batch 98.89 | loss  5.88 | ppl   358.89
| epoch   1 |  2000/ 2983 batches | lr 20.00 | ms/batch 98.63 | loss  5.85 | ppl   348.68
| epoch   1 |  2200/ 2983 batches | lr 20.00 | ms/batch 98.92 | loss  5.74 | ppl   311.96
| epoch   1 |  2400/ 2983 batches | lr 20.00 | ms/batch 98.78 | loss  5.75 | ppl   315.15
| epoch   1 |  2600/ 2983 batches | lr 20.00 | ms/batch 98.90 | loss  5.74 | ppl   310.26
| epoch   1 |  2800/ 2983 batches | lr 20.00 | ms/batch 99.65 | loss  5.62 | ppl   276.45
-----------------------------------------------------------------------------------------
| end of epoch   1 | time: 309.66s | valid loss  6.92 | valid ppl  1008.39
-----------------------------------------------------------------------------------------
=========================================================================================
| End of training | test loss  6.87 | test ppl   964.76
=========================================================================================
```

**Comment by paulkre** [comment]
Although, now that I look at it more closely, `valid ppl` and `test ppl` still seem to be much higher than on CUDA.

**Comment by rgb1380** [comment]
@qqaatw I have the same issue with the nightly build

**Comment by qqaatw** [comment]
> @qqaatw I have the same issue with the nightly build

Can you please open a new issue with a small repro? As most people have reported that their issues have been resolved. Thank you.

**Comment by qqaatw** [comment]
> > > @qqaatw I have the same issue with the nightly build
> 
> > 
> 
> > Can you please open a new issue with a small repro? As most people have reported that their issues have been resolved. Thank you.
> 
> 
> 
> @qqaatw 
> 
> @paulkre had already provide a minimally reproducible code but he is now reporting again that the results on CUDA and MPS are different. It is the quickest way to check this out. 

I would not expect the results from CUDA and MPS are the the same as the ways they implement LSTM likely differ.

Opening a new issue not only provides a better visibility for public but also lets us know your environment information, e.g. macOS version, for debugging purposes.

**Comment by Roberto-Candelario** [comment]
I'm having the same issue. I'm following a book tutorial. I am on torch version 2.4.1 using a M3 Max MacBook Pro 

I let the CPU run for a few hours to see it maybe I miss typed something but the CPU got better results. Almost identical to the tutorial. When I use 'mps' the results are completely different. I even downloaded the source code and still did not get the same results when using 'mps'.

**Comment by shaanchandra** [comment]
> I'm having the same issue. I'm following a book tutorial. I am on torch version 2.4.1 using a M3 Max MacBook Pro
> 
> I let the CPU run for a few hours to see it maybe I miss typed something but the CPU got better results. Almost identical to the tutorial. When I use 'mps' the results are completely different. I even downloaded the source code and still did not get the same results when using 'mps'.

Facing the same issue for a transformer based diffusion model. On CPU loss decreases, on MPS (just changing the device='mps') the loss increases for the exact same code. 

![Image](https://github.com/user-attachments/assets/dd0287a1-3369-422d-bc95-16c39abac818)
